### PR TITLE
Port color.cpp to Rust

### PR DESCRIPTION
**Pull request by Xiretza**
## Description

Ports the bit-packed `rgb_color_t` to Rust.

Does it makes sense to implement FFI here? I haven't found a way to convince cxx not to require a pointer indirection for every color value, which would probably defeat the whole size optimization.

### CODE REVIEW

File: fish-rust/src/common.rs [hunk 1]
```diff
@@ -115,3 +115,75 @@ pub fn valid_func_name(name: &wstr) -> bool {
...
+/// const COLORS: &[(&wstr, u32)] = &[
+///     // must be in alphabetical order
+///     (L!("blue"), 0x0000ff),
+///     (L!("green"), 0x00ff00),
+///     (L!("red"), 0xff0000),
+/// ];
+///
+/// assert_sorted_by_name!(COLORS, 0);
+/// ```
+macro_rules! assert_sorted_by_name {
```
Comment (ridiculousfish) [inline fish-rust/src/common.rs:139]: Whoa, I did not expect this to be possible in Rust. We may need to disable this check on older versions, but very cool!

File: fish-rust/src/color.rs [hunk 2]
```diff
@@ -0,0 +1,453 @@
...
+pub struct RgbColor {
+    #[bits = 3]
+    typ: Type,
+    #[bits = 5]
+    flags: Flags,
+    /// Either the R, G and B bytes for [`Type::Rgb`], or the name index for [`Type::Named`].
+    data: B24,
+}
+
+const _: () = assert!(size_of::<RgbColor>() <= 4, "RgbColor is too big");
```
Comment (ridiculousfish) [inline fish-rust/src/color.rs:66]: Thank you for faithfully propagating this assert, but let's remove it, this isn't important.

File: fish-rust/src/color.rs [hunk 3]
```diff
@@ -0,0 +1,453 @@
...
+
+#[bitfield]
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct RgbColor {
+    #[bits = 3]
+    typ: Type,
+    #[bits = 5]
+    flags: Flags,
+    /// Either the R, G and B bytes for [`Type::Rgb`], or the name index for [`Type::Named`].
+    data: B24,
```
Comment (ridiculousfish) [inline fish-rust/src/color.rs:63]: Let's either make this a Rust union, matching the C++; or if you are feeling ambitious go ahead and combine it with `typ` into an `enum`.

**Review by ridiculousfish — CHANGES_REQUESTED** [review]
This looks very solid - thank you, well done!

Let's not use bitfields here - my original use of bitfields was not well motivated and I regret it, and I prefer to not add the complexity of `modular_bitfield`. Please make these ordinary bools in an ordinary struct. That and a few other nits and it looks good. Thank you again!

**Comment by Xiretza** [comment]
Not having to do bitfields certainly makes it a lot easier! It's all regular Rust types now.

File: fish-rust/src/common.rs [hunk 4]
```diff
@@ -115,3 +115,75 @@ pub fn valid_func_name(name: &wstr) -> bool {
...
+
+            let mut i = 0;
+            let mut prev: Option<&wstr> = None;
+            while i < $slice.len() {
+                let cur = $slice[i].$field;
+                if let Some(prev) = prev {
+                    assert!(
+                        matches!(
+                            cmp_slice(prev.as_char_slice(), cur.as_char_slice()),
+                            Ordering::Less
```
Comment (mqudsi) [inline fish-rust/src/common.rs:174]: Do you want this to be strictly `Ordering::Less` or do we envision a case where we might actually want `!Ordering::Greater`?

Comment (Xiretza) [reply fish-rust/src/common.rs:174]: You're right, the C++ macro is `<=`. Fixed.

**Comment by ridiculousfish** [comment]
Merged as 7585ddf9262829ae00426150b2c52bfae2478d34, thank you!

**Pull request closed without merging**